ravenframework/TSA/TimeSeriesAnalyzer.py

```diff
@@ -101,23 +118,86 @@ def setDefaults(self, settings):
       settings['seed'] = None
     return settings
 
-  def getResidual(self, initial, params, pivot, settings):
+  def canAcceptMissingValues(self):
     """
-      Removes trained signal from data and find residual
-      @ In, initial, np.array, original signal shaped [pivotValues, targets], targets MUST be in
-                               same order as self.target
+      Checks if the algorithm can accept missing values (generally NaN).
+      @ Out, _acceptsMissingValues, bool, if the characterization algorithm accepts missing values
+    """
+    # NOTE Signals may have missing values, and this is incompatible with some algorithms. As
+    ## missing values will generally require special handling specific to the algorithm, this
+    ## behavior defaults to False. It is left to each algorithm to implement how these missing
+    ## values are handled.
+    return self._acceptsMissingValues
+
+  @abc.abstractmethod
+  def writeXML(self, writeTo, params):
+    """
+      Allows the engine to put whatever it wants into an XML to print to file.
+      @ In, writeTo, xmlUtils.StaticXmlElement, entity to write to
+      @ In, params, dict, parameters from training this ROM
+      @ Out, None
+    """
+
+
+class TimeSeriesGenerator(TimeSeriesAnalyzer):
+  """
+    Act as a mix-in class for algorithms that can generate synthetic time histories. This is
+    reserved exclusively for stochastic algorithms. Deterministic generative algorithms should NOT
+    inherit from this class.
```

Comment:
Several other comments you've made revolve around this distinction of generators needing to be stochastic or not, so I'll respond to that general idea here. My thinking there was that it can be helpful to know if an algorithm is stochastic or deterministic, and having that be difference of type was a handy way to make that distinction. I also noticed that deterministic functions tend to have a residual from fitting that we want to pass on to downstream algorithms, while generators (thinking about ARMA and GANs) often don't have that. Having stochastic and determinstic algorithms have different methods (from different base classes) would make sense in that case. In the end, the implementations for training and evaluation in TSAUser don't require there to be a stochastic generator in the list of algorithms, so maybe there's no concrete benefit from the distinction right now.

Reply:
I can see that it would be beneficial to have stochastic generators distinct. I would rather see `Generators` further subdivided into `Deterministic` and `Stochastic` if necessary, rather than make all deterministic `Generators` into `Transformers`. This is related to the discussion on where Fourier gets classified, above.

As to residuals, we actually do see a future where stochastic signals can subtract out the parts of the signal that are "likely" attributed to the noise characteristics, and residual pieces passed downstream. We haven't had a case to use this on yet, but it seems like it would be necessary to complete the toolset. I don't know how we'd go about a GAN informing us the likelihood a sample originated from its sampling, but for ARIMA it should be straightforward, I think.

Reply:
I'd be fine narrowing the scope of what is a transformer and adding some distinction for `Generators` between deterministic and stochastic generators. It seems that every algorithm would need `getResidual` and `getComposite` methods then, so we might not need the `Transformer` base class at all then, since those are the only unique methods that the class adds.

What would you think of adding those methods to the `TimeSeriesAnalyzer` base class, dropping the `Transformer` base class entirely, and making classes which are exclusively `Transformer`s right now (e.g. `LogTransformer`) direct subclasses of `TimeSeriesAnalyzer`? Then the distinction of stochastic vs deterministic generators couldbe easily handled with a class attribute/checker function within `Generators` since we probably don't want a different interface between the two.